### ISSUE TITLE
Need more basic explanations and examples for bitfield

### ISSUE THREAD

**Issue by ZhouHansen** [datrs/book#1]
These are some 🔦questions:

I red up [bitfield](https://github.com/fb55/bitfield), [sparse-bitfield](https://github.com/datrs/sparse-bitfield) though. I've still struggled to figure out the bitfield module for few days. The source of [indexed-field](https://github.com/mafintosh/indexed-bitfield) and [hypercore/bitfield](https://github.com/mafintosh/hypercore) are both hard to read. And [sleep.md/bitfiled](https://github.com/datproject/docs/blob/master/papers/sleep.md#bitfield) is similar to this document. So I decide to ask questions now.

1. Is this module for [sparse file](https://en.wikipedia.org/wiki/Sparse_file), or something similar to it？ 
2. In [Types of  Bitfields](https://github.com/datrs/book/blob/master/src/bitfield.md#types-of-bitfields)
"Data Bitfield: Indicates which data you have, and which data you don't. " What does 'the data you have' means? Does this means data entries I've already downloaded, then when update coming, I don't need to download them again? I don't know. Hypercore is a append-only log. So if there are 3 entries in `metadata.data`, the Data Bitfield will be `111000000`?

I want to figure out this one first.

**Comment by yoshuawuyts** [comment]
> What does 'the data you have' means? Does this means data entries I've already downloaded, then when update coming, I don't need to download them again?

Yeah, that covers the gist of it! -- because Hypercore is append-only, data can be tracked by just flipping a bit, and once it's flipped we don't need to download that piece of data ever again. `sparse-bitfield` is mostly an efficient way of storing that data on disk.

> So if there are 3 entries in metadata.data, the Data Bitfield will be 111000000?

The data is stored in a tree format, so it would be represented differently on disk. But it seems you've got the right intuition about it!

---

I'll be spending some time after this month to write more docs. Questions like these are super helpful to know where we should focus more on writing docs. Please keep the questions coming! :tada:

**Comment by ZhouHansen** [comment]
> ### Example
> ```txt
> bits:  00101
> index: 01234
>  ```
> The set above contains `2` and `4`.

Hypercore is append-only, I think when the previous bit is 0 then the next rest of bits should all be 0, so why this happens: the `0` and `1` is 0, but `2` is 1.

**Comment by yoshuawuyts** [comment]
Oh yeah, so while Hypercore is append-only, it also supports sparse replication. So:
- new data can only be appended
- data can be randomly replicated between peers

The bitfield represents the data that you have stored on your machine. Because we can do sparse copies (e.g. only copy a few entries in hypercore, or just one file in Hyperdrive) it must be able to represent that in its data structure, which means interior mutability.

Does that make sense?

**Comment by ZhouHansen** [comment]
Yeah, that's what I want to make sure. Thanks

**Comment by ZhouHansen** [comment]
Why does it means [interior mutability](https://doc.rust-lang.org/book/second-edition/ch15-05-interior-mutability.html)?

**Comment by yoshuawuyts** [comment]
> I'm learning hypercore, I think SLEEP files prefixed by content are in charge of hypercore, files prefixed by metadata are in charge of hyperdrive. Right?

Not quite; they're both `hypercore` instances. One keeps track of the content, the other keeps track of the metadata. In newer version of `hyperdrive`, the metadata core is a `hypercore` that's formatted as a [`hyperdb`](https://github.com/mafintosh/hyperdb).
